labs/lab03.md

```diff
@@ -0,0 +1,52 @@
+# Lab 3: Testing
+
+In this lab you will get familiar with different testing strategies and practice with writing good unit tests in Java and TypeScript.
+
+## Concepts
+**Structural testing vs Specification testing:** Two distinct types of unit testing are specification-based (historically: black-box) testing and structural (white-box) testing. Specification testing is when you write high-level test cases against the specification of a program, without using (or testing) the specific underlying implementation of the program. In structural testing, your tests are instead tailored to the specific underlying implementation of the code you are testing.
+
+**Code coverage:** Coverage is the amount of code that has been executed by the tests. There are several coverage metrics, but one common metric line coverage is simply the percentage of lines (or statements) that have been executed by your unit tests. 
+
+## Deliverables
+
+- [ ] Write tests for LinkedIntQueue in TypeScript and ArrayIntQueue in Java and execute them from within the IDE.
+- [ ] Achieve 100% line coverage for ArrayIntQueue and show coverage results in the IDE or as HTML report. 
+- [ ] Compare structural testing vs specification testing and analyze tradeoffs. 
```

Comment:
can we make this more concrete? what should students do? should they write a paragraph? should they explain something to the TA? should they answer a TA question?

I'm fine with making this about talking to the TA rather than writing, but I'd like to make this much more concrete. Maybe "Describe tradeoffs between structural testing and specification testing to the TA and demonstrate your understanding by answering follow up questions. (no written answer needed)"?

Reply:
sure that sounds good. when saying "by answering follow up questions" do we want a list of follow up questions or its fine for students just to Describe tradeoffs between structural testing and specification testing?

Reply:
I want them to demonstrate an understanding rather than just learn prepared statements. The idea is that they answer easy questions we ask them on the spot, not to prepare memorized statements for a list of possible questions.